src/sys/unix/uds.rs

```diff
@@ -0,0 +1,296 @@
+    let sockaddr = mem::MaybeUninit::<libc::sockaddr_un>::zeroed();
+
+    // This is safe to assume because a `libc::sockaddr_un` filled with `0`
+    // bytes is properly initialized.
+    //
+    // `0` is a valid value for `sockaddr_un::sun_family`; it is
+    // `libc::AF_UNSPEC`.
+    //
+    // `[0; 108]` is a valid value for `sockaddr_un::sun_path`; it begins an
+    // abstract path.
+    let mut sockaddr = unsafe { sockaddr.assume_init() };
+
+    sockaddr.sun_family = libc::AF_UNIX as libc::sa_family_t;
+
+    let bytes = path.as_os_str().as_bytes();
+    match (bytes.get(0), bytes.len().cmp(&sockaddr.sun_path.len())) {
+        // Abstract paths don't need a null terminator
+        (Some(&0), Ordering::Greater) => {
+            return Err(io::Error::new(
+                io::ErrorKind::InvalidInput,
+                "path must be no longer than SUN_LEN",
+            ));
+        }
+        (_, Ordering::Greater) | (_, Ordering::Equal) => {
+            return Err(io::Error::new(
+                io::ErrorKind::InvalidInput,
+                "path must be shorter than SUN_LEN",
+            ));
+        }
+        _ => {}
+    }
+
+    for (dst, src) in sockaddr.sun_path.iter_mut().zip(bytes.iter()) {
+        *dst = *src as libc::c_char;
+    }
+
+    let mut socklen = sun_path_offset() + bytes.len();
+    match bytes.get(0) {
+        // The struct has already been zeroes so the null byte for pathname
+        // addresses is already there.
+        Some(&0) | None => {}
+        Some(_) => socklen += 1,
+    }
+
+    Ok((sockaddr, socklen as libc::socklen_t))
+}
+
+fn pair_descriptors(mut fds: [i32; 2], flags: i32) -> io::Result<()> {
+    #[cfg(not(any(target_os = "ios", target_os = "macos", target_os = "solaris")))]
+    let flags = flags | libc::SOCK_NONBLOCK | libc::SOCK_CLOEXEC;
+
+    syscall!(socketpair(libc::AF_UNIX, flags, 0, fds.as_mut_ptr()))?;
+
+    // Darwin and Solaris don't have SOCK_NONBLOCK or SOCK_CLOEXEC.
+    //
+    // For platforms that don't support flags in `socket`, the flags must be
+    // set through `fcntl`. The `F_SETFL` command sets the `O_NONBLOCK` bit.
+    // The `F_SETFD` command sets the `FD_CLOEXEC` bit.
+    #[cfg(any(target_os = "ios", target_os = "macos", target_os = "solaris"))]
+    {
+        syscall!(fcntl(fds[0], libc::F_SETFL, libc::O_NONBLOCK))?;
+        syscall!(fcntl(fds[0], libc::F_SETFD, libc::FD_CLOEXEC))?;
+        syscall!(fcntl(fds[1], libc::F_SETFL, libc::O_NONBLOCK))?;
+        syscall!(fcntl(fds[1], libc::F_SETFD, libc::FD_CLOEXEC))?;
+    }
+    Ok(())
+}
+
+// On Linux, this funtion equates to the same value as
+// `size_of::<sun_path>()`, but some other implementations include other
+// fields before `sun_path`, so the expression more portably describes the
+// size of the address structure.
+fn sun_path_offset() -> usize {
+    let sockaddr = unsafe { mem::MaybeUninit::<libc::sockaddr_un>::uninit().assume_init() };
```

Comment:
Yes, this is undefined behavior. It is explicitly called out in https://doc.rust-lang.org/std/mem/union.MaybeUninit.html.

Reply:
Yep good call. I originally thought this was okay because `MaybeUninit<T>` guarantees size and alignment, but it specifies later:

>Rust does not in general guarantee that the fields of a Foo<T> have the same order as a Foo<U> even if T and U have the same size and alignment

Calculating an offset from this is therefore UB because the fields may not be ordered!

That being said-this can be removed by initializing the struct with what you pointed out above:

```
let mut sockaddr = libc::sockaddr_un {
    sun_family: libc::AF_UNIX as libc::sa_family_t,
    sun_path: [0; 108],
};
```